contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -282,22 +283,82 @@ contract MixinMatchOrders is
     )
         private
     {
-        bytes memory zrxAssetData = ZRX_ASSET_DATA;
-        // Order makers and taker
-        _dispatchTransferFrom(
-            leftOrderHash,
-            leftOrder.makerAssetData,
-            leftOrder.makerAddress,
-            rightOrder.makerAddress,
-            matchedFillResults.right.takerAssetFilledAmount
-        );
-        _dispatchTransferFrom(
-            rightOrderHash,
-            rightOrder.makerAssetData,
-            rightOrder.makerAddress,
-            leftOrder.makerAddress,
-            matchedFillResults.left.takerAssetFilledAmount
-        );
+        address leftFeeRecipientAddress = leftOrder.feeRecipientAddress;
+        address rightFeeRecipientAddress = rightOrder.feeRecipientAddress;
+
+        // Settle left order taker asset and maker fees.
+        if (
+            leftOrder.makerAddress == leftFeeRecipientAddress &&
+            leftOrder.makerFeeAssetData.equals(leftOrder.takerAssetData)
+        ) {
+            // Maker is fee recipient and the maker fee asset is the same as the taker asset.
+            // We can transfer the taker asset and maker fees in one go.
+            _dispatchTransferFrom(
```

Comment:
If I'm reading this line correctly, the Right Maker is paying their fee to the Left Fee Recipient. But it should go to the Right Fee Recipient.

This is the intended logic:
```
left.makerFeePaid → paid by Left Maker to Left Fee Recipient
left.takerFeePaid → paid by Taker to to Left Fee Recipient

right.makerFeePaid → paid by Right Maker to Right Fee Recipient
right.takerFeePaid → paid by Taker to Right Fee Recipient
```

Reply:
You're right, that is exactly what it's doing. I admit I have to work this out in my head every time I see it, but I think this behavior is correct.

We are basically collapsing the two transfers in the following `else` clause because the the `leftMaker == leftFeeRecipient` and the `leftMakerFeeAssetData == leftTakerAssetData`. So `rightMaker` can just pay the combined `leftTakerAsset` and `leftMakerFee` to `leftMaker`/`leftFeeRecipient` since it's all the same asset going to the same place.

Reply:
Spoke with @hysz and he pointed out there were some other fundamental errors in this settlement optimization (for one, takers were actually paying more than their fair share). This affected both `matchOrders()` and `fillOrders()`. The combinatorial tests also weren't covering these scenarios so they slipped under the radar. Totes my bad!

I just now updated things so the only optimizations we do now are avoiding maker fee transfers if `maker == feeRecipient` and taker fee transfers if `taker == feeRecipient`. There are now also explicit tests to cover all these (and the previous) scenarios.

Reply:
Combinatorial tests now take even longer though. :tired_face:
I am definitely adding these to demo sprint.